### PR TITLE
report: close drop down menu when focus is lost

### PR DESCRIPTION
**Pull request by johnemau**
**Summary**
#### Problem
The report drop down menu does not close when the escape key
is pressed in the devtools.

***Gif of bug: menu does not close when escape key is pressed and console opens***
devtools-menu-console-esc-bug

This creates a challenge for customer relying on the keyboard that wish
to dismiss the drop down.

This also does not align with the experience of the WAI ARIA practice
navigation menu button example [1].

#### Root cause
The escape key event is captured and not propagated by the
devtools to toggle the console drawer.

#### Solution
Close the menu when focus moves out of the menu.

This better aligns the menu with the Navigation menu button example for
which its interaction design is based on [1].

**Code Change**
* Add a menu focus out handler to close the drop down menu
* Unfortunately JSDOM does not support focusout events so JSDOM integration tests
  are not possible [2], instead I have added unit test for the onMenuFocusOut handler.
* A TypeScript upgrade is needed to better support the focusout event [3].

***Gif of bug fixed: menu does close when escape key is pressed and console opens***
devtools-menu-console-esc-bug-fixed

***Gif of menu escape key closing the menu***
devtools-menu-console-esc-bug-fixed-demo-esc

***Gif of menu tabbing closing the menu***
devtools-menu-console-esc-bug-fixed-demo-tabbing

**Related Issues/PRs**
#9433

[1]: https://www.w3.org/TR/2019/NOTE-wai-aria-practices-1.1-20190207/examples/menu-button/menu-button-links.html
[2]: https://github.com/jsdom/jsdom/issues/2708
[3]: https://github.com/microsoft/TypeScript/issues/30716

### CODE REVIEW

File: lighthouse-core/report/html/renderer/report-ui-features.js [hunk 1]
```diff
@@ -744,6 +747,21 @@ class DropDown {
...
   }
 
+  /**
+   * Focus out handler for the drop down menu.
+   * @param {Event} e
+   */
+  onMenuFocusOut(e) {
+    // The focusout event is not supported in our current version of typescript (3.5.3)
+    // https://github.com/microsoft/TypeScript/issues/30716
+    const focusEvent = /** @type {FocusEvent} */ (e);
```
Comment (johnemau) [inline lighthouse-core/report/html/renderer/report-ui-features.js]: Is there appetite to upgrade typescript?

**Comment by johnemau** [comment]
I am going to look into adding a unit test for onMenuFocusOut to improve code coverage per codecov results.

File: lighthouse-core/report/html/renderer/report-ui-features.js [hunk 2]
```diff
@@ -744,6 +747,21 @@ class DropDown {
...
+    const focusedEl = /** @type {?HTMLElement} */ (focusEvent.relatedTarget);
```
Comment (connorjclark) [inline lighthouse-core/report/html/renderer/report-ui-features.js]: is this nullable?

File: lighthouse-core/report/html/renderer/report-ui-features.js (hunk 1 above)
Comment (connorjclark) [reply lighthouse-core/report/html/renderer/report-ui-features.js]: I'm always up for upgrading TS, but we may not be at critical mass yet ... @patrickhulce what do you think?

File: lighthouse-core/report/html/renderer/report-ui-features.js (hunk 2 above)
Comment (johnemau) [reply lighthouse-core/report/html/renderer/report-ui-features.js]: Yes, relatedTarget might be null.
https://developer.mozilla.org/en-US/docs/Web/API/FocusEvent/FocusEvent

Comment (connorjclark) [reply lighthouse-core/report/html/renderer/report-ui-features.js]: ah, I see, this is another thing that newer TS has corrected.

**Comment by connorjclark** [comment]
> I have just finished writing a few unit tests and will add this to this PR but I am happy to follow up with UX tests if you think they would be valuable.

If you feel that covers the new behavior well enough that is fine. Was just suggesting an alternative in case unit tests weren't possible.

> I do not know about the Print preview issue, do you have a link to the issue or repro steps? I will take a look.

no crbug, just something I noticed. Printing doesn't seem to be a feature anyone uses so no big deal, was just curious. If you choose the Print option from the tool menu and don't see this tool menu in the Print results, then you've fixed it :)

**Comment by johnemau** [comment]
> If you feel that covers the new behavior well enough that is fine. Was just suggesting an alternative in case unit tests weren't possible.

I gotcha, the unit tests I added are good enough coverage.

> no crbug, just something I noticed. Printing doesn't seem to be a feature anyone uses so no big deal, was just curious. If you choose the Print option from the tool menu and don't see this tool menu in the Print results, then you've fixed it :)

Unfortunately this change did not fix Print preview :(

image

But this looks like a CSS issue, the menu is hidden by `visibility:hidden` and so I fixed it locally by adding:
 ```css
@media print { 
  .lh-tools__dropdown { 
      /* the drop down is not supported on paper ;) */
      display: none; 
  }
```

I will get a PR out with that fix shortly.

**Comment by johnemau** [comment]
> I will get a PR out with that fix shortly.

Done: #10216

File: lighthouse-core/test/report/html/renderer/report-ui-features-test.js [hunk 3]
```diff
@@ -395,5 +395,50 @@ describe('ReportUIFeatures', () => {
         assert.strictEqual(nextNode, nodes[2]);
       });
     });
+
+    describe('onMenuFocusOut', () => {
+      it('should toggle active class when focus relatedTarget is null', () => {
+        dropDown._toggleEl.click();
```
Comment (connorjclark) [inline lighthouse-core/test/report/html/renderer/report-ui-features-test.js]: nit: move these first two lines to `beforeEach` scoped to this new `describe`?

File: lighthouse-core/report/html/renderer/report-ui-features.js [hunk 4]
```diff
@@ -744,6 +747,21 @@ class DropDown {
     }
   }
 
+  /**
+   * Focus out handler for the drop down menu.
+   * @param {Event} e
+   */
+  onMenuFocusOut(e) {
+    // The focusout event is not supported in our current version of typescript (3.5.3)
```
Comment (connorjclark) [inline lighthouse-core/report/html/renderer/report-ui-features.js]:
```suggestion
    // TODO: The focusout event is not supported in our current version of typescript (3.5.3)
```

Let's just handle this later.

Comment (johnemau) [reply lighthouse-core/report/html/renderer/report-ui-features.js]: Done!

File: lighthouse-core/test/report/html/renderer/report-ui-features-test.js (hunk 3 above)
Comment (johnemau) [reply lighthouse-core/test/report/html/renderer/report-ui-features-test.js]: Good idea, thanks! Done.

Comment (johnemau) [reply lighthouse-core/test/report/html/renderer/report-ui-features-test.js]: Good idea, thanks! Done.

**Pull request merged**